### ISSUE TITLE
Sonic Pi is not listening on port 4557

### ISSUE THREAD

**Issue by Garciaj007** [Widdershin/sonic-pi-cli#22]
# Problem
```
sonic_pi play :A4
ERROR: Sonic Pi is not listening on 4557 - is it running?
```
## Description
when I run this command I get this error, I have sonic pi installed and open. Sonic Pi works fine, I have read about this issue/fix here, https://github.com/samaaron/sonic-pi/issues/1881#issuecomment-386576806

However, I think Sonic Pi wouldn't work at all, Right?
Doing some testing, is the UDP port the one that connects to the OSC Protocol inside Sonic Pi?

**Comment by rbnpi** [comment]
You need to update to the latest version which addresses this issue. Latest versions of Sonic Pi use a different randomly allocated port. Also the version of Sonic Pi 3.1 supplied weith Raspbian doesn't work as it does not have the necessary support built in. Instead use the later 3.2.2 version which you can download from http://r.newman.ch/rpi/sonic-pi-3.2.2/ That will work with the latest version of sonic-pi-cli

**Comment by Garciaj007** [comment]
Yep, so I have Sonic-Pi  v3.2.2, running, also this is on windows (might've forgotten to mention this). running `ruby 2.6.6p146 (2020-03-31 revision 67876) [x64-mingw32]`, `gem version 3.0.3` and I did the `gem install sonic-pi-cli` version  0.1.3 but still get the same error...

**Comment by rbnpi** [comment]
Hmm. I’ve just tried using it with sp3.2.2 installed on windows 10 here and it works fine. I wonder if you have relocated your .sonic-pi folder containing the log files? Sonic-pi-cli expects it to be in your user ENV[‘HOME’] folder. It retrieves the port to use from the log file C:\Users\IUSERNAME\log\server-output.log
If it can’t find it it assumes you are using an early version of Sonic Pi which used port 4557. If it can’t find that it gives up. Check that your log file is in the correct location. 
The port it searches for is in the line. Listen port:

**Comment by Garciaj007** [comment]
I am not sure if you mean that the server-output.log file should exist in C:\Users\USERNAME\.sonic-pi\log\server-output.log _*OR*_ C:\Users\IUSERNAME\log\server-output.log. I've tried doing it both ways and I still get the same error.

**Comment by rbnpi** [comment]
What is the full location path of .sonic-pi/log folder on your installation?

**Comment by Widdershin** [comment]
Is it possible to run Sonic Pi on a particular port? If not maybe we should just add support for that.

**Comment by rbnpi** [comment]
The port is randomly chosen by SP since version 3.1 I think Sam may include options later for a config file (see the comments in the commit below) letting the user set it. You can already hack the source to do that. I think SP on windows may cause problems as some users redirect the location of the .sonic-pi folder, particularly if they are working on a network. In this case environmental variables would n=have to be employed for sonic-pi-cli to select the correct location to get hold of the port. It is a bit of a hack to read the port from the log file, although I have found it satisfactory on my Mac, Windows PC and linux machines including Raspberry PI.
The system changes in Sonic Pi for the present port setup were implemented in this commit.
https://github.com/samaaron/sonic-pi/commit/d245d93c5b797ad76fa333f829c32d67480af96c#diff-f8b53271d5cf1e8d451989a9e81e1c4e

**Comment by jonnyhotchkiss** [comment]
> Sonic-pi-cli expects it to be in your user ENV[‘HOME’] folder. It retrieves the port to use from the log file C:\Users\IUSERNAME\log\server-output.log

Hi I've just found this interesting little app whilst perusing sp-cli stuff, and then this thread whilst encountering this issue.

Is it possible to provide the port?

In case it helps, I notice (on 4.4, on windows 10 home)

- 172.18.... local ip
- incoming osc port appears to be set (not sure if adjustable yet, via UI or commands or init.rb)
- %userprofile%\.sonic-pi\log has  **spider.log** - nothing seen for "server-output.log" - in case that helps?
- set (ENV) has `HOMEPATH`  - but I see `HOME` in Ruby / SonicPi ENV 👍 points to user directory 

Thanks

**Comment by rbnpi** [comment]
I have made sonic-pi-cli4 which works with versions after sp4.0 use command sonic_pi4 instead of sonic_pi. You can install with gem install sonic-pi-cli4. (May need sudo)

**Comment by jonnyhotchkiss** [comment]
Thanks! how do I call it? `sonic_pi`  calls the other one, tried `sonic_pi_cli4` 
I couldn't locate the repo, is it on GH?

**Comment by jonnyhotchkiss** [comment]
oops - got it! `sonic_pi4` 👍

**Comment by m-zielinski** [comment]
`ERROR: Sonic Pi is not listening on  - is it running?`

`sonic_pi_cli4` seems to ignore the port set in log/server-output.log

**Comment by rbnpi** [comment]
sonic_pi_cli4 gets the port information from the spider.log file. It is only applicable for Sonic Pi versions 4.0 and greater up to and including version 4.5 the current release.
It expects to find the file in the location ENV['HOME']+'/.sonic-pi/log/spider.log
where ENV['HOME'] is the users home directory.

PREVIOUS versions of Sonic PI used data from the server-output.log but these work with sonic_pi_cli which is what should be installed for them. You can have both versions installed, as long as you use sonic_pi_cli for earlier versions of Sonic PI and sonic_pi_cli4 for versions >= 4.0 including the current release version 4.5

**Comment by farvardin** [comment]
@m-zielinski @rbnpi I got the same error with sonic-pi 4.6 (which I've built myself from sources), I have a spider.log in my home folder (on linux) but it doesn't contain the port number in it (4560 for me). 

I've changed your sonic_piRBN.rb file, line 10 to hardcode the port used on my system:

``@port=4560``

now sonic-pi doesn't complain about the port, but I don't get any sound either...

Too bad this didn't happen:

" Official command line support is planned for Sonic Pi 3.0."

**Comment by rbnpi** [comment]
sonic_pi4 only works for versions 4 to 4.5 A different mechanism is needed to extract the ports required for version 4.6
I have an updated gem version named sonic-pi-cli5 which works with this but I haven't published it yet. It will use sonic_pi5 instead of sonic_pi4 to send OSC commands.
All of these sonic-pi-cli versions are using unofficial methods to produce the required gem file, and it gets quite messy with effectively 4 differnt version required depending on the version of Sonic Pi you are using it with. It would become too complex to try and produce a unified version. I'll try and get something published soon, probably just on a fork on my own site https//:github/rbnpi but I'm quite busy at present.
EDIT
by the way port 4560 is NOT used for any cli commands. A randomly generated port number is used for processing osc commands like /run-code The later versions of sonic-pi-cli obrain suhc information form the sonic-pi log files. The format of these has changed in version 4.6.
port 4560 IS used for sending commands which the user decodes in a running sonic-pi buffer. EG you might send the message "/setcutoff", 60  to port 4560, and write code in your program to respond to this message being received.
